File: Python3/0220. 存在重复元素 III.py
class Solution:
    def containsNearbyAlmostDuplicate(self, nums: List[int], k: int, t: int) -> bool:

        def bucket_id(n):
           return n // (t + 1)
        
        bucket = dict()

        for i, n in enumerate(nums):
            if len(bucket) > k :
                del bucket[bucket_id(nums[i - k - 1])]

            bid = bucket_id(n)

            if bucket.get(bid) != None:
                return True
            bucket[bid] = n

            if bucket.get(bid - 1) != None and abs(bucket[bid - 1] - n) <= t:
                return True
            
            if bucket.get(bid + 1) != None and abs(bucket[bid + 1] - n) <= t:
                return True
        
        return False

File: C++/0220. 存在重复元素 III.cpp
class Solution {
public:
    int bucketId(long n, long t) 
    {
        return n < 0 ? -(-n / (t + 1) + 1) : n / (t + 1);
    }

    bool containsNearbyAlmostDuplicate(vector<int>& nums, int k, int t) {        
        map<int, int> bucket;
        long n;
        int id;
        for(int i=0;i<nums.size();++i) {
            n = nums[i]; 

            if(bucket.size() > k) {
                bucket.erase(bucketId(nums[i - k - 1], t));
            }

            id = bucketId(n, t);

            if(bucket.count(id) != 0) {
                return true;
            }

            bucket[id] = n;
            
            if(bucket.count(id - 1) != 0 && abs(n - bucket[id - 1]) <= t ) {
                return true;
            }

            if(bucket.count(id + 1) != 0 && abs(n - bucket[id + 1]) <= t ) {
                return true;
            }
        }
        return false;

    }
};
